README.md
# Parquet Contrib

Parquet Contrib contains stuffs I build when I can't find any implementation out there for Parquet format.

## Features

* Apache Pig MultiStorage-esque

## Apache Pig

### MultiStorage-esque
A Storer (ParquetMultiStorer) that mimics the functionality of [PiggyBank MultiStorage](http://pig.apache.org/docs/r0.12.1/api/org/apache/pig/piggybank/storage/MultiStorage.html) are provided for storing data into Parquet in Pig in multi output format.

#### Build

Apologize for not having it in MavenRepository right now. Please be gentle, this is the first time I use mvn to package (have always been a Gradle guy)

To build the jars:
```
mvn package
```

#### Usage and Example

```
REGISTER '/path/to/parquet-pig-bundle-1.5.0.jar';
REGISTER '/path/to/parquet-hadoop-contrib-1.1.1.jar'; -- generated from the above build step
REGISTER '/path/to/parquet-pig-contrib-1.1.1.jar';    -- generated from the above build step

SET parquet.compression snappy

data = LOAD 'raw' USING PigStorage() as (bucket, name, value);
data = ORDER data BY bucket;
STORE data INTO '/path/to/output' USING parquet.pig.ParquetMultiStorer('0');
```

**(Required) '/path/to/output'** the path where output directories and files will be created

**(Required) '0'**: Index of the field whose values should be used to create directories and parquet files (field 'bucket' in this case)

Let **'gold'** and **'silver'** be the unique values of field 'bucket'. Then output will look like this:

```
/path/to/output/gold.snappy.parquet/gold-00000-1111111111111.snappy.parquet
/path/to/output/gold.snappy.parquet/gold-00000-1415613091563.snappy.parquet
/path/to/output/gold.snappy.parquet/gold-00001-1111111111111.snappy.parquet
...
/path/to/output/silver.snappy.parquet/silver-00000-1111111111111.snappy.parquet
/path/to/output/silver.snappy.parquet/silver-00001-1234567891234.snappy.parquet
```

The affix **'0000\*'** is the task-id of the mapper/reducer task executing this store.
The affix **'1111111111111'** is the timestamp of the file being created (new in version 1.1.0).

#### Caveats
* In version 1.1.1, there can be only active writer at all time. So it is imperative that you sort the result by the key column. This is digusting yet I can't find out why the heck that if there is more than 1 writer, the output files are corrupted and the buckets are filled in incorrectly. Also the previous implemented cache is useless now. Sigh. <del>Internally ParquetMultiStorer creates a Writer for each key, so if you have a large number of buckets (number of unique keys), you winds up creating way too many filehandles all at the same time. It will be very slow or most of the time, you will encounter Out of Memory. Workaround's in Version 1.1.0 in which it uses Guava LoadingCache to control the number of file handles can be opened.</del>
* Missing unit test gives poor confidence (I do have a few rudimentary ones in Pig and Bash, but doing them in Java is a pain though).
* Astute reader might notice that the contrib package has the same location as the official package (parquet.hadoop), this is because it uses *CodecFactory.BytesCompressor* which is private. Probably a better way will be [Create your own objects](https://github.com/apache/incubator-parquet-mr#create-your-own-objects).
* If you are MultiStorage user, the prefix/task-id might look differently depending on your locale. MultiStorage uses *NumberFormat* to format the task-id (which is influenced by system locale), while I use *String.format* with zero-padded instead to remove this inconsistency.
* I used the code from parquet hadoop *1.5.0*, thus if you are using CHD5 or PivotalHD, you will need to replace the *pig-bundle-1.2.x* jar. WriterVersion property is missing in CHD5 parquet io column, while TupleWriteSupport.init is missing in PivotalHD.

## License

Licensed under the Apache License, Version 2.0: http://www.apache.org/licenses/LICENSE-2.0


CHANGES.md
### Version 1.1.1 ### 
* Fix corrupted output by setting only 1 active writer at all time

### Version 1.1.0 ### 
* Use Guava LoadingCache to control the maximum number of open file handles,
  thus avoids the Heap error.
* Add 'maxNumberOfWriters' parameter to dictate how many open file handles can
  be created.
* Pig/ParquetMultiStorer now accepts a 2nd parameter 'maxNumberOfWriters'
  (default to '10')

### Version 1.0.0 ### 


parquet-hadoop-contrib/src/main/java/parquet/hadoop/ParquetMultiOutputFormat.java
/**
 * Copyright 2012 Twitter, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package parquet.hadoop;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.mapred.JobConf;
import org.apache.hadoop.mapreduce.Job;
import org.apache.hadoop.mapreduce.JobContext;
import org.apache.hadoop.mapreduce.OutputCommitter;
import org.apache.hadoop.mapreduce.RecordWriter;
import org.apache.hadoop.mapreduce.TaskAttemptContext;
import org.apache.hadoop.mapreduce.TaskID;
import org.apache.hadoop.mapreduce.lib.output.FileOutputFormat;

import java.io.IOException;

import parquet.Log;
import parquet.column.ParquetProperties.WriterVersion;
import parquet.hadoop.api.WriteSupport;
import parquet.hadoop.api.WriteSupport.WriteContext;
import parquet.hadoop.codec.CodecConfig;
import parquet.hadoop.metadata.CompressionCodecName;
import parquet.hadoop.util.ConfigurationUtil;

import static parquet.Log.INFO;
import static parquet.Preconditions.checkNotNull;
import static parquet.hadoop.ParquetWriter.DEFAULT_BLOCK_SIZE;
import static parquet.hadoop.ParquetWriter.DEFAULT_PAGE_SIZE;
import static parquet.hadoop.util.ContextUtil.getConfiguration;

/**
 * OutputFormat to write to a Parquet file
 * <p/>
 * It requires a {@link WriteSupport} to convert the actual records to the underlying format.
 * It requires the schema of the incoming records. (provided by the write support)
 * It allows storing extra metadata in the footer (for example: for schema compatibility purpose when converting from a different schema language).
 * <p/>
 * The format configuration settings in the job configuration:
 * <pre>
 * # The block size is the size of a row group being buffered in memory
 * # this limits the memory usage when writing
 * # Larger values will improve the IO when reading but consume more memory when writing
 * parquet.block.size=134217728 # in bytes, default = 128 * 1024 * 1024
 *
 * # The page size is for compression. When reading, each page can be decompressed independently.
 * # A block is composed of pages. The page is the smallest unit that must be read fully to access a single record.
 * # If this value is too small, the compression will deteriorate
 * parquet.page.size=1048576 # in bytes, default = 1 * 1024 * 1024
 *
 * # There is one dictionary page per column per row group when dictionary encoding is used.
 * # The dictionary page size works like the page size but for dictionary
 * parquet.dictionary.page.size=1048576 # in bytes, default = 1 * 1024 * 1024
 *
 * # The compression algorithm used to compress pages
 * parquet.compression=UNCOMPRESSED # one of: UNCOMPRESSED, SNAPPY, GZIP, LZO. Default: UNCOMPRESSED. Supersedes mapred.output.compress*
 *
 * # The write support class to convert the records written to the OutputFormat into the events accepted by the record consumer
 * # Usually provided by a specific ParquetOutputFormat subclass
 * parquet.write.support.class= # fully qualified name
 *
 * # To enable/disable dictionary encoding
 * parquet.enable.dictionary=true # false to disable dictionary encoding
 *
 * # To enable/disable summary metadata aggregation at the end of a MR job
 * # The default is true (enabled)
 * parquet.enable.summary-metadata=true # false to disable summary aggregation
 * </pre>
 * <p/>
 * If parquet.compression is not set, the following properties are checked (FileOutputFormat behavior).
 * Note that we explicitely disallow custom Codecs
 * <pre>
 * mapred.output.compress=true
 * mapred.output.compression.codec=org.apache.hadoop.io.compress.SomeCodec # the codec must be one of Snappy, GZip or LZO
 * </pre>
 * <p/>
 * if none of those is set the data is uncompressed.
 *
 * @param <K> the key to be split by
 * @param <T> the type of the materialized records
 * @author Julien Le Dem, Nguyen Duc Hoa
 */
public class ParquetMultiOutputFormat<K, T> extends FileOutputFormat<K, T> {

    public static final String BLOCK_SIZE = "parquet.block.size";
    public static final String PAGE_SIZE = "parquet.page.size";
    public static final String COMPRESSION = "parquet.compression";
    public static final String WRITE_SUPPORT_CLASS = "parquet.write.support.class";
    public static final String DICTIONARY_PAGE_SIZE = "parquet.dictionary.page.size";
    public static final String ENABLE_DICTIONARY = "parquet.enable.dictionary";
    public static final String VALIDATION = "parquet.validation";
    public static final String WRITER_VERSION = "parquet.writer.version";
    public static final String ENABLE_JOB_SUMMARY = "parquet.enable.summary-metadata";
    private static final Log LOG = Log.getLog(ParquetMultiOutputFormat.class);
    private static final int MAX_NO_OF_WRITERS = 10;

    private WriteSupport<T> writeSupport;
    private ParquetMultiOutputCommitter committer;
    private int maxNumberOfWriters;

    /**
     * constructor used when this OutputFormat in wrapped in another one (In Pig for example)
     *
     * @param writeSupportClass  the class used to convert the incoming records
     * @param schema             the schema of the records
     * @param extraMetaData      extra meta data to be stored in the footer of the file
     * @param maxNumberOfWriters max number of open file handles
     */
    public <S extends WriteSupport<T>> ParquetMultiOutputFormat(S writeSupport, int maxNumberOfWriters) {
        this.writeSupport = writeSupport;
        this.maxNumberOfWriters = maxNumberOfWriters;
    }


    /**
     * constructor used when this OutputFormat in wrapped in another one (In Pig for example)
     *
     * @param writeSupportClass the class used to convert the incoming records
     * @param schema            the schema of the records
     * @param extraMetaData     extra meta data to be stored in the footer of the file
     */
    public <S extends WriteSupport<T>> ParquetMultiOutputFormat(S writeSupport) {
        this.writeSupport = writeSupport;
        this.maxNumberOfWriters = MAX_NO_OF_WRITERS;
    }

    /**
     * used when directly using the output format and configuring the write support implementation
     * using parquet.write.support.class
     */
    public <S extends WriteSupport<T>> ParquetMultiOutputFormat() {
        this.maxNumberOfWriters = MAX_NO_OF_WRITERS;
    }

    public static void setWriteSupportClass(Job job, Class<?> writeSupportClass) {
        getConfiguration(job).set(WRITE_SUPPORT_CLASS, writeSupportClass.getName());
    }

    public static void setWriteSupportClass(JobConf job, Class<?> writeSupportClass) {
        job.set(WRITE_SUPPORT_CLASS, writeSupportClass.getName());
    }

    public static Class<?> getWriteSupportClass(Configuration configuration) {
        final String className = configuration.get(WRITE_SUPPORT_CLASS);
        if (className == null) {
            return null;
        }
        final Class<?> writeSupportClass = ConfigurationUtil.getClassFromConfig(configuration, WRITE_SUPPORT_CLASS, WriteSupport.class);
        return writeSupportClass;
    }

    public static void setBlockSize(Job job, int blockSize) {
        getConfiguration(job).setInt(BLOCK_SIZE, blockSize);
    }

    public static void setPageSize(Job job, int pageSize) {
        getConfiguration(job).setInt(PAGE_SIZE, pageSize);
    }

    public static void setDictionaryPageSize(Job job, int pageSize) {
        getConfiguration(job).setInt(DICTIONARY_PAGE_SIZE, pageSize);
    }

    public static void setCompression(Job job, CompressionCodecName compression) {
        getConfiguration(job).set(COMPRESSION, compression.name());
    }

    public static void setEnableDictionary(Job job, boolean enableDictionary) {
        getConfiguration(job).setBoolean(ENABLE_DICTIONARY, enableDictionary);
    }

    public static boolean getEnableDictionary(JobContext jobContext) {
        return getEnableDictionary(getConfiguration(jobContext));
    }

    public static int getBlockSize(JobContext jobContext) {
        return getBlockSize(getConfiguration(jobContext));
    }

    public static int getPageSize(JobContext jobContext) {
        return getPageSize(getConfiguration(jobContext));
    }

    public static int getDictionaryPageSize(JobContext jobContext) {
        return getDictionaryPageSize(getConfiguration(jobContext));
    }

    public static CompressionCodecName getCompression(JobContext jobContext) {
        return getCompression(getConfiguration(jobContext));
    }

    public static boolean isCompressionSet(JobContext jobContext) {
        return isCompressionSet(getConfiguration(jobContext));
    }

    public static void setValidation(JobContext jobContext, boolean validating) {
        setValidation(getConfiguration(jobContext), validating);
    }

    public static boolean getValidation(JobContext jobContext) {
        return getValidation(getConfiguration(jobContext));
    }

    public static boolean getEnableDictionary(Configuration configuration) {
        return configuration.getBoolean(ENABLE_DICTIONARY, true);
    }

    public static int getBlockSize(Configuration configuration) {
        return configuration.getInt(BLOCK_SIZE, DEFAULT_BLOCK_SIZE);
    }

    public static int getPageSize(Configuration configuration) {
        return configuration.getInt(PAGE_SIZE, DEFAULT_PAGE_SIZE);
    }

    public static int getDictionaryPageSize(Configuration configuration) {
        return configuration.getInt(DICTIONARY_PAGE_SIZE, DEFAULT_PAGE_SIZE);
    }

    public static WriterVersion getWriterVersion(Configuration configuration) {
        String writerVersion = configuration.get(WRITER_VERSION, WriterVersion.PARQUET_1_0.toString());
        return WriterVersion.fromString(writerVersion);
    }

    public static CompressionCodecName getCompression(Configuration configuration) {
        return CodecConfig.getParquetCompressionCodec(configuration);
    }

    public static boolean isCompressionSet(Configuration configuration) {
        return CodecConfig.isParquetCompressionSet(configuration);
    }

    public static void setValidation(Configuration configuration, boolean validating) {
        configuration.setBoolean(VALIDATION, validating);
    }

    public static boolean getValidation(Configuration configuration) {
        return configuration.getBoolean(VALIDATION, false);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public RecordWriter<K, T> getRecordWriter(TaskAttemptContext taskAttemptContext)
            throws IOException, InterruptedException {
        final Configuration conf = getConfiguration(taskAttemptContext);

        CompressionCodecName codec = getCodec(taskAttemptContext);
        String extension = codec.getExtension() + ".parquet";
        TaskID taskId = taskAttemptContext.getTaskAttemptID().getTaskID();

        Path workPath = ((ParquetMultiOutputCommitter)
                getOutputCommitter(taskAttemptContext)).getWorkPath();

        return getRecordWriter(
                conf, workPath, extension, String.format("%05d", taskId.getId()), codec);
    }

    public RecordWriter<K, T> getRecordWriter(Configuration conf, Path workPath, String extension, String taskId, CompressionCodecName codec)
            throws IOException, InterruptedException {
        final WriteSupport<T> writeSupport = getWriteSupport(conf);

        CodecFactory codecFactory = new CodecFactory(conf);
        int blockSize = getBlockSize(conf);
        if (INFO) LOG.info("Parquet block size to " + blockSize);
        int pageSize = getPageSize(conf);
        if (INFO) LOG.info("Parquet page size to " + pageSize);
        int dictionaryPageSize = getDictionaryPageSize(conf);
        if (INFO) LOG.info("Parquet dictionary page size to " + dictionaryPageSize);
        boolean enableDictionary = getEnableDictionary(conf);
        if (INFO) LOG.info("Dictionary is " + (enableDictionary ? "on" : "off"));
        boolean validating = getValidation(conf);
        if (INFO) LOG.info("Validation is " + (validating ? "on" : "off"));
        WriterVersion writerVersion = getWriterVersion(conf);
        if (INFO) LOG.info("Writer version is: " + writerVersion);

        WriteContext init = writeSupport.init(conf);

        return new ParquetMultiRecordWriter<K, T>(
                workPath,
                extension,
                taskId,
                conf,
                writeSupport,
                init.getSchema(),
                init.getExtraMetaData(),
                blockSize, pageSize,
                codecFactory.getCompressor(codec, pageSize),
                dictionaryPageSize,
                enableDictionary,
                validating,
                writerVersion,
                getMaxNumberOfWriters());
    }

    /**
     * @param configuration to find the configuration for the write support class
     * @return the configured write support
     */
    @SuppressWarnings("unchecked")
    public WriteSupport<T> getWriteSupport(Configuration configuration) {
        if (writeSupport != null) return writeSupport;
        Class<?> writeSupportClass = getWriteSupportClass(configuration);
        try {
            return (WriteSupport<T>) checkNotNull(writeSupportClass, "writeSupportClass").newInstance();
        } catch (InstantiationException e) {
            throw new BadConfigurationException("could not instantiate write support class: " + writeSupportClass, e);
        } catch (IllegalAccessException e) {
            throw new BadConfigurationException("could not instantiate write support class: " + writeSupportClass, e);
        }
    }

    public int getMaxNumberOfWriters() {
        return this.maxNumberOfWriters;
    }

    @Override
    public OutputCommitter getOutputCommitter(TaskAttemptContext context)
            throws IOException {
        if (committer == null) {
            Path output = getOutputPath(context);
            committer = new ParquetMultiOutputCommitter(output, context);
        }
        return committer;
    }

    private CompressionCodecName getCodec(TaskAttemptContext taskAttemptContext) {
        return CodecConfig.from(taskAttemptContext).getCodec();
    }

}


parquet-hadoop-contrib/src/main/java/parquet/hadoop/ParquetMultiRecordWriter.java
/**
 * Copyright 2012 Twitter, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package parquet.hadoop;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.mapreduce.RecordWriter;
import org.apache.hadoop.mapreduce.TaskAttemptContext;

import java.io.IOException;
import java.util.concurrent.ExecutionException;
import java.util.HashMap;
import java.util.Map;

import com.google.common.cache.Cache;
import com.google.common.cache.CacheBuilder;
import com.google.common.cache.CacheLoader;
import com.google.common.cache.LoadingCache;
import com.google.common.cache.RemovalListener;
import com.google.common.cache.RemovalNotification;

import parquet.column.ParquetProperties.WriterVersion;
import parquet.hadoop.CodecFactory.BytesCompressor;
import parquet.hadoop.ParquetFileWriter;
import parquet.hadoop.ParquetRecordWriter;
import parquet.hadoop.api.WriteSupport;
import parquet.schema.MessageType;

/**
 * Writes records to a Parquet file
 *
 * @param <K> the key to be split by
 * @param <T> the type of the materialized records
 * @author Julien Le Dem, Nguyen Duc Hoa
 * @see ParquetMultiOutputFormat
 */
public class ParquetMultiRecordWriter<K, T> extends RecordWriter<K, T> {

    private Path workPath;
    private String extension;
    private String taskId;
    private Configuration conf;
    private WriteSupport<T> writeSupport;
    private MessageType schema;
    private Map<String, String> extraMetaData;
    private int blockSize;
    private int pageSize;
    private BytesCompressor compressor;
    private int dictionaryPageSize;
    private boolean enableDictionary;
    private boolean validating;
    private WriterVersion writerVersion;
    private int maxNumberOfWriters;

    private LoadingCache<K, ParquetRecordWriter> cache;

    /**
     * @param workPath           the path to the output directory (temporary)
     * @param extension          the codec extension + .parquet
     * @param taskId             the zero-padded task ID
     * @param writeSupport       the class to convert incoming records
     * @param schema             the schema of the records
     * @param extraMetaData      extra meta data to write in the footer of the file
     * @param blockSize          the size of a block in the file (this will be approximate)
     * @param compressor         the compressor used to compress the pages
     * @param dictionaryPageSize the threshold for dictionary size
     * @param enableDictionary   to enable the dictionary
     * @param validating         if schema validation should be turned on
     * @param maxNumberOfWriters max number of open file handles
     */
    public ParquetMultiRecordWriter(
            Path workPath,
            String extension,
            String taskId,
            Configuration conf,
            WriteSupport<T> writeSupport,
            MessageType schema,
            Map<String, String> extraMetaData,
            int blockSize, int pageSize,
            BytesCompressor compressor,
            int dictionaryPageSize,
            boolean enableDictionary,
            boolean validating,
            WriterVersion writerVersion,
            int maxNumberOfWriters) {
        this.workPath = workPath;
        this.extension = extension;
        this.taskId = taskId;
        this.conf = conf;
        this.writeSupport = writeSupport;
        this.schema = schema;
        this.extraMetaData = extraMetaData;
        this.blockSize = blockSize;
        this.pageSize = pageSize;
        this.compressor = compressor = compressor;
        this.dictionaryPageSize = dictionaryPageSize;
        this.enableDictionary = enableDictionary;
        this.validating = validating;
        this.writerVersion = writerVersion;
        this.maxNumberOfWriters = maxNumberOfWriters;
    }

    /**
     * {@inheritDoc}
     */
    @SuppressWarnings("rawtypes")
    @Override
    public void close(TaskAttemptContext context) throws IOException, InterruptedException {
        this.cache.invalidateAll();
    }

    public LoadingCache<K, ParquetRecordWriter> getCache() {
        if (this.cache == null) {
            final String extension = this.getExtension();
            final String taskId = this.getTaskId();
            final Path workPath = this.getWorkPath();
            final Configuration conf = this.getConf();
            final MessageType schema = this.getSchema();
            final WriteSupport<T> writeSupport = this.getWriteSupport();
            final Map<String, String> extraMetaData = this.getExtraMetaData();
            final int blockSize = this.getBlockSize();
            final int pageSize = this.getPageSize();
            final BytesCompressor compressor = this.getCompressor();
            final int dictionaryPageSize = this.getDictionaryPageSize();
            final boolean enableDictionary = this.isEnableDictionary();
            final boolean validating = this.isValidating();
            final WriterVersion writerVersion = this.getWriterVersion();

            CacheLoader<K, ParquetRecordWriter> loader =
                    new CacheLoader<K, ParquetRecordWriter> () {
                public ParquetRecordWriter load(K key) throws Exception {
                    final String fieldValue = key.toString();
                    final long timestamp = System.currentTimeMillis();

                    Path path = new Path(fieldValue + extension, fieldValue + '-'
                            + taskId + '-' + timestamp + extension);
                    Path file = new Path(workPath, path);

                    ParquetFileWriter fw = new ParquetFileWriter(
                            conf,
                            schema,
                            file);
                    fw.start();

                    return new ParquetRecordWriter<T>(
                            fw,
                            writeSupport,
                            schema,
                            extraMetaData,
                            blockSize,
                            pageSize,
                            compressor,
                            dictionaryPageSize,
                            enableDictionary,
                            validating,
                            writerVersion);
                }
            };

            RemovalListener<K, ParquetRecordWriter> removalListener =
                    new RemovalListener<K, ParquetRecordWriter>() {
                public void onRemoval(RemovalNotification<K, ParquetRecordWriter> removal) {
                    ParquetRecordWriter writerToRemove = removal.getValue();
                    try {
                        writerToRemove.close(null);
                    } catch (IOException ioe) {
                        throw new RuntimeException("Exception on closing cached writer", ioe);
                    } catch (InterruptedException ite) {
                        throw new RuntimeException(ite);
                    }
                }
            };

            this.cache = CacheBuilder.newBuilder()
                .maximumSize(getMaxNumberOfWriters())
                .removalListener(removalListener)
                .build(loader);
        }

        return this.cache;
    }

    /**
     * {@inheritDoc}
     */
    @SuppressWarnings("unchecked")
    @Override
    public void write(K key, T value) throws IOException, InterruptedException {
        try {
            // WARNING:
            // This is digusting yet I can't find out why the heck that if
            // there are 2 writers, the output files are corrupted and the
            // buckets are filled in incorrectly. Also the cache is useless
            // now since only 1 writer will be active at all time.
            if (key != getCurrentKey()) {
                if (this.cache != null)
                    getCache().invalidate(getCurrentKey());
                setCurrentKey(key);
            }
            getCache().get(key).write((Void) null, value);
        } catch (ExecutionException ee) {
            throw new RuntimeException("Exception on getting cached writer", ee);
        }
    }

    public Path getWorkPath() {
        return workPath;
    }

    public Configuration getConf() {
        return conf;
    }

    public String getExtension() {
        return extension;
    }

    public String getTaskId() {
        return taskId;
    }

    public WriteSupport<T> getWriteSupport() {
        return writeSupport;
    }

    public MessageType getSchema() {
        return schema;
    }

    public Map<String, String> getExtraMetaData() {
        return extraMetaData;
    }

    public int getBlockSize() {
        return blockSize;
    }

    public int getPageSize() {
        return pageSize;
    }

    public BytesCompressor getCompressor() {
        return compressor;
    }

    public int getDictionaryPageSize() {
        return dictionaryPageSize;
    }

    public boolean isEnableDictionary() {
        return enableDictionary;
    }

    public boolean isValidating() {
        return validating;
    }

    public WriterVersion getWriterVersion() {
        return writerVersion;
    }

    public int getMaxNumberOfWriters() {
        return maxNumberOfWriters;
    }

    private K currentKey;

    public K getCurrentKey() {
        return this.currentKey;
    }

    public void setCurrentKey(K currentKey) {
        this.currentKey = currentKey;
    }
}


parquet-pig-contrib/src/main/java/parquet/pig/ParquetMultiStorer.java
/**
 * Copyright 2012 Twitter, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package parquet.pig;

import java.io.IOException;
import java.util.Properties;

import org.apache.hadoop.fs.Path;
import org.apache.hadoop.mapreduce.Job;
import org.apache.hadoop.mapreduce.OutputFormat;
import org.apache.hadoop.mapreduce.RecordWriter;
import org.apache.hadoop.mapreduce.lib.output.FileOutputFormat;
import org.apache.pig.ResourceSchema;
import org.apache.pig.ResourceStatistics;
import org.apache.pig.StoreFunc;
import org.apache.pig.StoreMetadata;
import org.apache.pig.data.Tuple;
import org.apache.pig.impl.logicalLayer.schema.Schema;
import org.apache.pig.impl.util.UDFContext;
import org.apache.pig.impl.util.Utils;
import org.apache.pig.parser.ParserException;

import parquet.io.ParquetEncodingException;
import parquet.pig.TupleWriteSupport;
import parquet.hadoop.ParquetMultiOutputFormat;

/**
 * A pig storer for the Parquet file format for splitting the output
 * data into multiple directories and files dynamically based on
 * user specified key field in the output tuple.
 *
 * Sample Usage:
 * <code>
 * data = LOAD 'raw' USING PigStorage() as (bucket, name, value);
 * STORE data INTO '/path/to/output' USING parquet.pig.ParquetMultiStorer('0');
 * </code>
 * 
 * (Required) '/path/to/output' the path where output directories and files will be created
 * (Required) '0': Index of the field whose values should be used to create 
 * directories and parquet files (field 'bucket' in this case)
 *
 * Let 'gold' and 'silver' be the unique values of field 'bucket'. Then output will
 * look like this:
 *
 * /path/to/output/gold.snappy.parquet/gold-00000.snappy.parquet
 * /path/to/output/gold.snappy.parquet/gold-00001.snappy.parquet
 * /path/to/output/gold.snappy.parquet/gold-00002.snappy.parquet
 * ...
 * /path/to/output/silver.snappy.parquet/silver-00000.snappy.parquet
 * /path/to/output/silver.snappy.parquet/silver-00001.snappy.parquet
 *
 * The prefix '0000*' is the task-id of the mapper/reducer task executing
 * this store.
 *
 * see {@link ParquetMultiOutputFormat} for available parameters.
 * <p/>
 * It uses a TupleWriteSupport to write Tuples into the ParquetOutputFormat
 * The Pig schema is automatically converted to the Parquet schema using {@link PigSchemaConverter}
 * and stored in the file
 *
 * @author Julien Le Dem, Nguyen Duc Hoa
 */
public class ParquetMultiStorer extends StoreFunc implements StoreMetadata {

    private static final String SCHEMA = "schema";
    private static final int MAX_NO_OF_WRITERS = 10;

    private RecordWriter<String, Tuple> recordWriter;

    private String signature;

    private int splitFieldIndex = 0;
    private int maxNumberOfWriters;

    public ParquetMultiStorer(String splitFieldIndex)
            throws IOException {
        this(splitFieldIndex, Integer.toString(MAX_NO_OF_WRITERS));
    }

    public ParquetMultiStorer(String splitFieldIndex, String maxNumberOfWriters)
            throws IOException {
        this.splitFieldIndex = Integer.parseInt(splitFieldIndex);
        this.maxNumberOfWriters = Integer.parseInt(maxNumberOfWriters);

        if (this.maxNumberOfWriters <= 0)
            throw new IOException("Max Number of Writers must be larger than 0");
    }

    private Properties getProperties() {
        UDFContext udfc = UDFContext.getUDFContext();
        Properties p =
                udfc.getUDFProperties(this.getClass(), new String[]{signature});
        return p;
    }

    private Schema getSchema() {
        try {
            final String schemaString = getProperties().getProperty(SCHEMA);
            if (schemaString == null) {
                throw new ParquetEncodingException("Can not store relation in Parquet as the schema is unknown");
            }
            return Utils.getSchemaFromString(schemaString);
        } catch (ParserException e) {
            throw new ParquetEncodingException("can not get schema from context", e);
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void setStoreFuncUDFContextSignature(String signature) {
        super.setStoreFuncUDFContextSignature(signature);
        this.signature = signature;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void checkSchema(ResourceSchema s) throws IOException {
        getProperties().setProperty(SCHEMA, s.toString());
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public OutputFormat<String, Tuple> getOutputFormat() throws IOException {
        Schema pigSchema = getSchema();
        return new ParquetMultiOutputFormat<String, Tuple>(new TupleWriteSupport(pigSchema), getMaxNumberOfWriters());
    }

    /**
     * {@inheritDoc}
     */
    @SuppressWarnings({"rawtypes", "unchecked"}) // that's how the base class is defined
    @Override
    public void prepareToWrite(RecordWriter recordWriter) throws IOException {
        this.recordWriter = recordWriter;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void putNext(Tuple tuple) throws IOException {
        int splitFieldIndex = getSplitFieldIndex();

        if (tuple.size() <= splitFieldIndex) {
            throw new IOException("split field index:" + splitFieldIndex
                    + " >= tuple size:" + tuple.size());
        }

        try {
            Object field = tuple.get(splitFieldIndex);
            this.recordWriter.write(String.valueOf(field), tuple);
        } catch (InterruptedException e) {
            Thread.interrupted();
            throw new ParquetEncodingException("Interrupted while writing", e);
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void setStoreLocation(String location, Job job) throws IOException {
        FileOutputFormat.setOutputPath(job, new Path(location));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void storeSchema(ResourceSchema schema, String location, Job job)
            throws IOException {
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void storeStatistics(ResourceStatistics resourceStatistics, String location, Job job)
            throws IOException {
    }

    public int getSplitFieldIndex() {
        return splitFieldIndex;
    }

    public int getMaxNumberOfWriters() {
        return maxNumberOfWriters;
    }
}
